Use StackWalker API to get the caller class

marschall opened the pull request:
Make jaxb-core a Multi-Release JAR and use the StackWalker API in
Java 9+ to get the caller class in Utils#getClassLogger. This avoids
throwing and catching exceptions to get the caller class.

marschall commented:
We have exception tracing on in production. We notice quite a few exceptions cause by `org.glassfish.jaxb.core.Utils#getClassLogger()`. Java 9+ now offers the `java.lang.StackWalker` API to get the caller class more efficiently without throwing and catching an exception. A [Multi-Release JAR File](https://openjdk.java.net/jeps/238) can be used to use this API only in Java 9+.

marschall commented:
> In general, the fix as such looks good, thanks.
> 
> few drawbacks your solution have:
> 
> * ECA is not signed, without that, the PR cannot be accepted, see [the guideline](https://github.com/eclipse-ee4j/jaxb-ri/blob/master/CONTRIBUTING.md)

I signed the ECA but with a different email address. I don't see how an additonal email address can be added to my existing account so the easiest is probably if I create and additional account.

> * instead of copy & pasting & updating existing class, I'd rather see the new code being introduced through some static method in a helper class, see similar change in [istack-commons](https://github.com/eclipse-ee4j/jaxb-istack-commons/pull/41/files) or the pattern with package private MUtils/MrUtils used in other related projects, even in [jaxb-ri itself](https://github.com/eclipse-ee4j/jaxb-ri/blob/master/jaxb-ri/runtime/impl/src/main/java-mr/9/org/glassfish/jaxb/runtime/v2/MUtils.java)

I can certainly do this.

> * changes in pom.xml are not enough for the new class to be propagated to repackaged artifact (and thus to the distribution archive), compare your changes with [runtime](https://github.com/eclipse-ee4j/jaxb-ri/blob/master/jaxb-ri/runtime/impl/pom.xml) and its [repackaged version](https://github.com/eclipse-ee4j/jaxb-ri/blob/master/jaxb-ri/bundles/runtime/pom.xml) to see what and how needs to be done

If I change the approach to a static helper class then this would no longer be necessary, right?

lukasj commented:
wrt ECA & email - one should be able to add additional email to the github account under Settings -> Emails, email in sign-off should match the email in your eclipse.org account

wrt pom.xml changes - they are still necessary since repackaging also fully rebuilds the code. In short, the source bundle must contain META-INF/versions folder with versioned class too. The other module should be able to take it from there

marschall commented:
> wrt ECA & email - one should be able to add additional email to the github account under Settings -> Emails, email in sign-off should match the email in your eclipse.org account

Turns out I already had an account for that email, but the ECA was expired so I renewed it.

> wrt pom.xml changes - they are still necessary since repackaging also fully rebuilds the code. In short, the source bundle must contain META-INF/versions folder with versioned class too. The other module should be able to take it from there

Got it. I changed the `pom.xml` to be more similar to the one from runtime and added a unit test. I hope I'll have time to look into the repackaged artifact soon.

One thing that is unclear to me is the copyright headers. I just copied and pasted the Oracle headers. Is this correct or should this now be Eclipse?

marschall commented:
@lukasj I believe I have implemented all the requested changes.

lukasj commented:
yup, I only see no reason to have bundle and jar plugins doing the same thing... but I'l fix that myslef later...

marschall commented:
Thank you!
